modules/wfa-qt/src/indigo_api_callback_dut.c

```diff
@@ -98,21 +102,80 @@ void register_apis(void)
 #endif /*End Of CONFIG_P2P */
 }
 
+static int run_qt_command(const char *cmd)
+{
+	char buffer[64] = { 0 }, response[16] = { 0 };
+	size_t resp_len = sizeof(response);
+	int ret = 0;
+
+	indigo_logger(LOG_LEVEL_ERROR, "Running cmd: %s", cmd);
+	snprintf(buffer, sizeof(buffer), "%s", cmd);
+	if (ret < 0 || ret >= (int)sizeof(buffer)) {
+		indigo_logger(LOG_LEVEL_ERROR, "Failed to execute command");
+		goto done;
+	}
+
+	if (ctrl_conn) {
+		ret = wpa_ctrl_request(ctrl_conn, buffer, sizeof(buffer),
+					response, &resp_len, NULL);
+		if (ret) {
+			indigo_logger(LOG_LEVEL_ERROR,
+				      "Failed to send command to supplicant:%d", ret);
+			goto done;
+		}
+		if (resp_len == 0) {
+			indigo_logger(LOG_LEVEL_ERROR, "Failed to execute command");
+			goto done;
+		}
+
+		if (strncmp(buffer, "ADD_CRED", strlen("ADD_CRED")) == 0) {
+			cred_id = atoi(response);
+		} else if ((strstr(buffer, "WPS_AP_PIN")) ||
+			   (strstr(buffer, "WPS_PIN")) ||
+			    (strstr(buffer, "WPS_PBC"))) {
+			memset(wsc_pin, 0, sizeof(wsc_pin));
+			if (strncpy(wsc_pin, response, sizeof(wsc_pin)) == NULL) {
+				goto done;
+			}
+		} else if (strncmp(buffer, "STATUS", strlen("STATUS") == 0)) {
+			memset(status_res, 0, sizeof(status_res));
+			if (strncpy(status_res, response, sizeof(status_res)) == NULL) {
+				goto done;
+			}
+		}
+
+	}
+	indigo_logger(LOG_LEVEL_DEBUG, "Response: %s", response);
+	return 0;
+done:
+	return -1;
+}
+
+#define CHECK_RET(ret_value) \
+	do { \
+		if (ret_value < 0) { \
+			indigo_logger(LOG_LEVEL_ERROR, "Error occurred"); \
+			goto done; \
+		} \
+	} while (0)
+
 static int get_control_app_handler(struct packet_wrapper *req, struct packet_wrapper *resp)
 {
 	char ipAddress[INET_ADDRSTRLEN];
 	char buffer[S_BUFFER_LEN];
+	int ret = 0;
 #ifdef _VERSION_
-	snprintf(buffer, sizeof(buffer), "%s", _VERSION_);
+	CHECK_SNPRINTF(buffer, sizeof(buffer), ret, "%s", _VERSION_);
 #else
-	snprintf(buffer, sizeof(buffer), "%s", TLV_VALUE_APP_VERSION);
+	CHECK_SNPRINTF(buffer, sizeof(buffer), ret, "%s", TLV_VALUE_APP_VERSION);
 #endif
 	if (tool_addr) {
 		inet_ntop(AF_INET, &(tool_addr->sin_addr), ipAddress, INET_ADDRSTRLEN);
 		indigo_logger(LOG_LEVEL_DEBUG,
 			      "Tool Control IP address on DUT network path: %s", ipAddress);
 	}
 
+done:
 	fill_wrapper_message_hdr(resp, API_CMD_RESPONSE, req->hdr.seq);
 	fill_wrapper_tlv_byte(resp, TLV_STATUS, TLV_VALUE_STATUS_OK);
 	fill_wrapper_tlv_bytes(resp, TLV_MESSAGE, strlen(TLV_VALUE_OK), TLV_VALUE_OK);
```

Comment:
We might as well de-duplicate this part with something like this to start with, and I still see potential to make this a lot simpler, but this is a good starting point.

```
static int fill_resp_to_ctrl(struct packet_wrapper *req, struct packet_wrapper *resp, int status, char *message)
{
	fill_wrapper_message_hdr(resp, API_CMD_RESPONSE, req->hdr.seq);
	fill_wrapper_tlv_byte(resp, TLV_STATUS, status);
	fill_wrapper_tlv_bytes(resp, TLV_MESSAGE, strlen(message), message);

	return 0;
}

#define FILL_RESP_SUCCESS fill_resp_to_ctrl(req, resp, TLV_VALUE_STATUS_OK, TLV_VALUE_OK)
#define FILL_RESP_FAIL(message) fill_resp_to_ctrl(req, resp, TLV_VALUE_STATUS_NOT_OK, message)


static int send_sta_reconnect_handler(struct packet_wrapper *req, struct packet_wrapper *resp)
{
	CHECK_RET(run_qt_command("RECONNECT"));
	return FILL_RESP_SUCCESS;
err:
	return FILL_RESP_FAIL(TLV_VALUE_WPA_S_RECONNECT_NOT_OK);
}
```

Reply:
Should I do this now?